libethash-cl/CLMiner.cpp

```diff
@@ -361,7 +360,7 @@ void CLMiner::workLoop()
 			// Report results while the kernel is running.
 			// It takes some time because ethash must be re-evaluated on CPU.
 			if (nonce != 0) {
-				Result r = EthashAux::eval(current.seed, current.header, nonce);
+				Result r = EthashAux::eval(current.epoch, current.header, nonce);
```

Comment:
The only issue I have with this approach, is that you're further delaying the submission of the share by forcing a search of an unordered map to convert the epoch back to the seed value in eval. Would be better to store a local copy of the seed when the epoch changes and pass it to eval as required. Same for CLMIner.

Reply:
I changed the map from `seed => light_cache` to `epoch_number => light_cache` inside eval(). However, I need to do additional map lookup do recover epoch number from seed hash when receiving a new job from a pool.

Reply:
You could optimize toEpoch to cache the last result and reuse it if the call is for the same epoch, as it will be most of the time.

Reply:
Same optimization could apply to toEpoch.

Reply:
I plan to do it in https://github.com/chfast/ethash.

Reply:
How about I only keep single cached result instead of map?

Reply:
Sure, better yet for Ethminer app.

Reply:
I'm working on this improvement here.

Moreover, is the lock still needed if the toEpoch is only called from stratum tread?